docs/design-docs/03-pipeline-merge.md

```diff
@@ -0,0 +1,628 @@
+upstream and existing setter values are updated locally. This is the most common 
+use case where kpt fails to merge them. Since `name` field is not specified, `image`
+value is used to identify and merge the function
+
+```
+Original
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.0.1
+```
+```
+Updated upstream
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.0.1
+        new-setter: new-setter-value // new setter is added
+```
+```
+Local
+```
+
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.2.0 // value of tag is updated
+```
+```
+Current Output
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.0.1 // entire pipeline is overridden by upstream 
+        new-setter: new-setter-value
+```
+```
+Expected Output
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.2.0 // updated tag is preserved
+        new-setter: new-setter-value // new setter is pulled
+```
+
+In the above scenario, the value of the setter tag is not updated upstream, so 
+the modified local value will be preserved. But in the following example, both 
+upstream and local values change. So, similar to merging resources, upstream 
+value wins if the same fields in both upstream and local are updated.
+
+```
+Original
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/set-labels:v0.1
+      configPath: labels.yaml
+```
+```
+Updated upstream
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/set-labels:v0.1
+      configPath: labels-updated.yaml // upstream value changed
+```
+```
+Local
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/set-labels:v0.1
+      configPath: labels-local.yaml // local value changed
+```
+```
+Expected Output
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/set-labels:v0.1
+      configPath: labels-updated.yaml // upstream overrides local 
+```
+
+Similarly, the upstream version wins if both upstream and local are updated, else local is preserved.
+
+```
+Original
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/set-annotations:v0.1
+      configPath: annotations.yaml
+```
+```
+Updated upstream
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/set-annotations:v0.2.0
+      configPath: annotations.yaml
+```
+```
+Local
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/set-annotations:v0.1.1
+      configPath: annotations.yaml
+```
+```
+Expected Output
+```
+```yaml
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/set-annotations:v0.2.0
+      configPath: annotations.yaml
+```
```

Comment:
This probably answers my question about the key when we use `image` field. Looks like we are excluding the version.

Haven't thought through completely, but final version of function and final merged input value (especially when configMap is used inline) can result in function-input and function to be incompatible in the end.

Reply:
Could be. But it is good to have incompatible output so that user can correct it on the local.